npep/npep-133.md

```diff
@@ -0,0 +1,89 @@
+# NPEP-133: FQDN Selector for Egress Traffic
+
+* Issue:
+  [#133](https://github.com/kubernetes-sigs/network-policy-api/issues/133)
+* Status: Provisional
+
+## TLDR
+
+This enhancement proposes adding a new selector to specify egress peers using
+Fully Qualified Domain Names (FQDNs). 
+
+## Goals
+
+* Provide a selector to specify egress peers using a Fully Qualified Domain Name
+  (for example `kubernetes.io`).
+* Support a restricted set of regex matching capabilities when specifying FQDNs. 
+
+## Non-Goals
+
+* This enhancement does not include a FQDN selector for allowing ingress
+  traffic.
+* This enhancement does not include any L7 matching or filtering capabilities,
+  like matching HTTP traffic or URL paths.
+    * This selector should not control what DNS records are resolvable from a
+      particular workload.
+* This enhancement does not provide a mechanism for selecting in-cluster
+  endpoints using FQDNs. This is explicitly disallowed by the spec.
+    * To select Pods, Nodes, API Server, AdminNetworkPolicy has more first party
+      selector with better UX.
+* This enhancement does not specify the details of how traffic is routed to the
+  specified destination. For example, it does not prescribe details around NAT
+  or egress gateways.
+* This enhancement does not require a particular mechanism for securing DNS
+  resolution (e.g. DNSSEC or DNS-over-TLS).
+    * TODO: Is this up to implementations? Maybe an optional enum describing
+      what security mechanims to use and providers can chose whther to support
+      or not.
+
+## Introduction
+
+FQDN-based egress controls are a common enterprise security practice.
```

Comment:
[citation required]

People _want_ this functionality, but do they already have it outside of Kubernetes? You certainly can't do it with iptables/nftables. (You can say "`iptables -d kubernetes.io -j DROP`" but that means "resolve the IP addresses of `kubernetes.io` right now and create one static rule for each IP address it returns".)

Do external firewall products implement this? If so, how _exactly_ do they implement it, since if that's what users expect, then we should just do the same thing.

Reply:
I think there's a ton of prior art both inside and outside Kubernetes on some flavor of FQDN firewall.

- https://learn.microsoft.com/en-us/azure/firewall/fqdn-filtering-network-rules
- https://cloud.google.com/firewall/docs/firewall-policies-rule-details#domain-names-for-firewall
- https://docs.vmware.com/en/VMware-NSX/4.1/administration/GUID-63262728-CA72-47D2-8E4F-16617B63A9A4.html
- https://techdocs.f5.com/kb/en-us/products/big-ip-afm/manuals/product/network-firewall-policies-implementations-13-1-0/3.html
- https://docs.tigera.io/calico-enterprise/latest/network-policy/domain-based-policy
- https://docs.cilium.io/en/stable/security/policy/language/#dns-based
- https://docs.openshift.com/container-platform/4.13/networking/openshift_sdn/configuring-egress-firewall.html

As far as matching user expectation, I'm mostly focused on what existing K8s CNIs are providing as that's the real target demographic here. I will do some reading and definitely the goal of the NPEP is to find the least common denominator behaviors. I'm curious if you have a particular behavior you're wary of.

Reply:
OpenShift's current implementation (based on proactively resolving hostnames in FQDN rules and re-resolving when the TTL expires) has turned out to not work well at all for hosts behind load balancers/CDNs (which seem to be a lot of the hosts that people want to make rules involving).

We're probably moving to an approach where we intercept client DNS requests and update the rules based on the actual IPs returned to the client, but this isn't implemented yet, and it's unknown if it will actually work better.

Reply:
OK, it looks like:

||Azure|GCP|VMware|F5|Calico|Cilium|OpenShift (current)|OpenShift (future)|
|-|-|-|-|-|-|-|-|-|
|Implementation| DNS snooping | async DNS | DNS snooping | DNS snooping? | DNS snooping | DNS snooping | async DNS | DNS snooping |
|Wildcards?| :heavy_check_mark: | :x: | :heavy_check_mark: | :x: | :heavy_check_mark: | :heavy_check_mark: | :x: | :heavy_check_mark: |
|Egress rules? | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: |
Ingress rules? | :x: | :heavy_check_mark:| :x:? |:heavy_check_mark: | :x: | :x: | :x: | :x: |
Allow rules? | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: |
Deny rules? | :heavy_check_mark:? | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :x:? | :x: | :heavy_check_mark: | :x:? |

So...

- There seems to be a strong consensus for DNS snooping, wildcards, and "allow egress" rules.
- Most systems have "deny egress" rules.
- Most systems don't have ingress FQDN rules.
- GCP and (current) OpenShift, the only two that (definitely) use async DNS rather than DNS snooping, both explicitly warn in their docs that they don't work well with frequently-changing DNS names. None of the DNS snooping systems have such warnings.
  - GCP: "For highly dynamic domain names, there might be a discrepancy between the resolved IP addresses in the local cache and the resolved IP addresses in the firewall policy rule. In such a scenario, the egress firewall rule works reliably only when the VM uses the resolved IP addresses configured in the firewall policy rules." ... "Don't use ingress rules that use FQDN objects if the domain name resolution results are highly variable or domain name resolution uses a form of DNS-based load balancing."
  - OpenShift: "Because the egress firewall controller and pods asynchronously poll the same local name server, the pod might obtain the updated IP address before the egress controller does, which causes a race condition. Due to this current limitation, domain name usage in EgressNetworkPolicy objects is only recommended for domains with infrequent IP address changes."

Reply:
yup, go ahead